java/client/src/main/java/glide/api/models/GlideString.java

```diff
@@ -0,0 +1,90 @@
+/** Copyright GLIDE-for-Redis Project Contributors - SPDX Identifier: Apache-2.0 */
+package glide.api.models;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.concurrent.atomic.AtomicBoolean;
+import lombok.Getter;
+
+// TODO docs for the god of docs
+public class GlideString {
+
+    @Getter private byte[] bytes;
+    private String string = null;
+
+    /** Flag whether possibility to convert to string was checked. */
+    private final AtomicBoolean conversionChecked = new AtomicBoolean(false);
+
+    private GlideString() {}
+
+    public static GlideString of(String string) {
+        var res = new GlideString();
+        res.string = string;
+        res.bytes = string.getBytes(StandardCharsets.UTF_8);
+        return res;
+    }
+
+    public static GlideString of(byte[] bytes) {
+        var res = new GlideString();
+        res.bytes = bytes;
+        return res;
+    }
+
+    @Override
+    public String toString() {
+        return getString();
+    }
+
+    public String getString() {
+        if (string != null) {
+            return string;
+        }
+
+        assert canConvertToString() : "Value cannot be represented as a string";
+        return string;
+    }
+
+    public boolean canConvertToString() {
+        if (string != null) {
+            return true;
+        }
+
+        // double-checked locking
+        if (conversionChecked.get()) {
+            return false;
+        } else {
+            synchronized (this) {
+                if (conversionChecked.get()) {
+                    return false;
+                } else
+                    try {
+                        // TODO find a better way to check this
+                        // Detect whether `bytes` could be represented by a `String` without data corruption
+                        var tmpStr = new String(bytes, StandardCharsets.UTF_8);
```

Comment:
The conversion + check are both `O(n)` (i.e. very expensive)

Reply:
See here: http://unicode.org/mail-arch/unicode-ml/y2003-m02/att-0467/01-The_Algorithm_to_Valide_an_UTF-8_String

In particular, the section: "A state machine implementation"